torch/csrc/jit/passes/utils/subgraph_utils.cpp

```diff
@@ -233,68 +253,77 @@ void mergeNodeIntoSubgraph(
   }
 
   for (auto input : orderedClosedValues) {
-    if (inputsMap.count(input) == 0) {
+    if (externalValuesMap.count(input) == 0) {
       // Clone constants inside the subgraph instead of referencing them, to
       // enable more optimizations
       if (auto value = toIValue(input)) {
         auto nv = subgraph->insertConstant(*value);
         nv->copyMetadata(input);
-        inputsMap[input] = nv;
+        externalValuesMap[input] = nv;
       } else {
         // The common case: this is a regular input, so just register it with
         // the group node and inner subgraph
         subgraphNode->addInput(input);
         auto inputToGraph = subgraph->addInput();
         inputToGraph->copyMetadata(input);
-        inputsMap[input] = inputToGraph;
+        externalValuesMap[input] = inputToGraph;
       }
     }
   }
 
   // Merge the node into the graph
-  auto mergedNode = subgraph->insertNode(
-      subgraph->createClone(toMerge, [&](Value* v) { return inputsMap[v]; }));
-
-  // If n's outputs were inputs to `group`, remove them since we just merged
-  // n in.
-  //
-  // i.e.,
-  // x = f(w); group(x, y, z) becomes group(w, y, z).
-  // x, y, z = f(w); group(x, y, z) becomes group(w).
-  auto inputs = subgraphNode->inputs();
-  for (size_t i = 0; i < toMerge->outputs().size(); ++i) {
-    auto it = std::find(inputs.begin(), inputs.end(), toMerge->outputs()[i]);
-    if (it != inputs.end()) {
-      size_t p = it - inputs.begin();
-      subgraphNode->removeInput(p);
-      subgraph->inputs()[p]->replaceAllUsesWith(mergedNode->outputs()[i]);
-      subgraph->eraseInput(p);
+  auto mergedNode = subgraph->insertNode(subgraph->createClone(
+      toMerge, [&](Value* v) { return externalValuesMap[v]; }));
+
+  if (!merging_node_after_subgraph) {
+    // If n's outputs were inputs to `group`, remove them since we just merged
+    // n in.
+    //
+    // i.e.,
+    // x = f(w); group(x, y, z) becomes group(w, y, z).
+    // x, y, z = f(w); group(x, y, z) becomes group(w).
+    auto inputs = subgraphNode->inputs();
+    for (size_t i = 0; i < toMerge->outputs().size(); ++i) {
+      auto it = std::find(inputs.begin(), inputs.end(), toMerge->outputs()[i]);
+      if (it != inputs.end()) {
```

Comment:
I understand you didn't change this code but do you know when we can have `it == inputs.end()`?

Reply:
> If n's outputs were inputs to `group`, remove them since we just merged 

If any of `n`'s outputs weren't inputs to the subgraph.